### PR TITLE
Improve long values in dict literals

### LINKED ISSUE #620: Indent in dictionnary
Operating system: Linux CentOS 7
Python version: 3.6.6
*Black* version: 18.9b0
Does also happen on master: Yes ([Example on master](https://black.now.sh/?version=master&state=_Td6WFoAAATm1rRGAgAhARYAAAB0L-Wj4ADqAJhdAD2IimZxl1N_WlbvK5V8IfU9pv4fHdsviW3P8OT-SNb5Xz4fF9wjrFurCdlaGQF69cdr_HeUViGKVVrrM9Rz1hoddpYGgepUWcCIo6M347RwgNi05tlJraGsxgRe--DyTGLl8NHBOCIXrmsiR1D2esFMJsZxaAKOz7j9RjnB96QvwRWTfjRbkQOQzwv07jJs0a1-y7UsN4mmADQSfEZD9p8xAAG0AesBAAAAANJ_scRn-wIAAAAABFla))

I have this code (line length is greater than allowed)
```python
my_dict = {
    'a key in my dict': deformation_rupture_mean * constraint_rupture_mean / 100.0
}
```

So Black wrap the line, but without indent :
```python
my_dict = {
     "a key in my dict": deformation_rupture_mean
     * constraint_rupture_mean
     / 100.0
}
```
I find this indent not very readable.

If I add paranthesis, I get the code below, nice formatted in my opinion.
```python
my_dict = {
	"a key in my dict": (
		deformation_rupture_mean * constraint_rupture_mean / 100.0
	)
}
```
Should not Black add paranthesis in this case?

### LINKED ISSUE #808: Line too long: Lambda in dictionary value
Input (which I also think is a good format for output):
```python
{
    "some very long dictionary key string goes here":
        lambda: some_very_long_function_name(x, y, z)
}
```
Output:
```python
{
    "some very long dictionary key string goes here": lambda: some_very_long_function_name(
        x, y, z
    )
}
```
Note that the second line is >88 characters.

Possibly related to https://github.com/ambv/black/issues/781?

Operating system: Linux
Python version: 3.6.3
*Black* version: 19.3b0
Does also happen on master: yes

### LINKED ISSUE #3117: Black --preview mode produces invalid code in the first pass, causing second pass to fail
**Describe the bug**

Black's --preview mode produces invalid code in the first pass, causing second pass to fail.

**To Reproduce**

For example, take this code:

```python
some_dict = {
    "something_something":
        r"Lorem ipsum dolor sit amet, an sed convenire eloquentiam \t"
        r"signiferumque, duo ea vocibus consetetur scriptorem. Facer \t",
}
```

And run it with these arguments:

```sh
$ black --preview file.py
```

The resulting error is:

```
error: cannot format filepy: Cannot parse: 5:4:     r"signiferumque, duo ea vocibus consetetur scriptorem. Facer \t",

Oh no! 💥 💔 💥
1 file failed to reformat.
```

**Expected behavior**

It should succeed.

**Environment**

- Black's version: black, 22.3.1.dev43+gf51e537 (compiled: no)
- OS and Python version: Linux, Python (CPython) 3.9.12

**Additional context**

1. Looks like the first [`_format_str_once` call](https://github.com/psf/black/blob/f51e53726b39a177355a7917c91c56f390dda7ef/src/black/__init__.py#L1163) formats the code to:

```
some_dict = {
    "something_something": (
        r"Lorem ipsum dolor sit amet, an sed convenire eloquentiam \t"
    )
    r"signiferumque, duo ea vocibus consetetur scriptorem. Facer \t",
}
```
2. It's only reproducible with r-strings.

### PR DESCRIPTION
**Pull request by yilei**
### Description

This PR changes the preview style:

- Wraps long values in dict literals in parentheses. Fixes #620. Fixes #808.
- Correspondingly, parentheses around short values in dict literals are removed.
- In preview string processing, long lambda string values are now wrapped in parentheses.
- Wrapping dict values happens to fix #3117 where implicitly concatenated r-string dict values crash in string processing. Fixes #3117

Usually I'd like to keep the PR's scope minimal, but the lambda change & dict value change are depended by each other so it's hard to split into two PRs.

### Checklist - did you ...

- [x] Add an entry in `CHANGES.md` if necessary?
- [x] Add / update tests if necessary?
- [x] Add new / update outdated documentation?

### CODE REVIEW

**Comment by JelleZijlstra** [comment]
Notes from diff-shades:

```
             queryset = queryset.extra(
                 select={
                     "_prefetch_related_val_%s"
-                    % f.attname: "%s.%s"
-                    % (qn(join_table), qn(f.column))
+                    % f.attname: "%s.%s" % (qn(join_table), qn(f.column))
                     for f in fk.local_related_fields
                 }
             )
```
Note that both the key and the value in the dictcomp use `%` formatting. Looks a little weird, maybe we should put the key and value on separate lines. Seems out of scope for this PR though.

Pandas has a similarly odd-looking case with a `+` in the key:
```
         tests = [
             (
                 BusinessHour(),
                 {
                     Timestamp("2014-07-04 15:00")
-                    + Nano(5): Timestamp("2014-07-04 16:00")
-                    + Nano(5),
+                    + Nano(5): Timestamp("2014-07-04 16:00") + Nano(5),
                     Timestamp("2014-07-04 16:00")
-                    + Nano(5): Timestamp("2014-07-07 09:00")
-                    + Nano(5),
+                    + Nano(5): Timestamp("2014-07-07 09:00") + Nano(5),
                     Timestamp("2014-07-04 16:00")
-                    - Nano(5): Timestamp("2014-07-04 17:00")
-                    - Nano(5),
+                    - Nano(5): Timestamp("2014-07-04 17:00") - Nano(5),
                 },
             ),
```

Otherwise I like the changes from the PR: they make it easier to see when one dict key ends and the next one begins, because values get indented.

File: src/black/trans.py [hunk 1]
```diff
@@ -1951,6 +1954,25 @@ def do_transform(self, line: Line, string_idx: int) -> Iterator[TResult[Line]]:
                     f" (left_leaves={left_leaves}, right_leaves={right_leaves})"
                 )
                 old_rpar_leaf = right_leaves.pop()
+            elif right_leaves and right_leaves[-1].type == token.RPAR:
+                # Special case for lambda expressions as dict's value, e.g.:
+                #     my_dict = {
+                #        "key": lambda x: f"formatted: {x},
+                #     }
+                # After wrapping the dict's value with parentheses, the string is
+                # followed by a RPAR but it's opening bracket is lambda's, not
```
Comment (JelleZijlstra) [inline src/black/trans.py:1963]:
```suggestion
                # followed by a RPAR but its opening bracket is lambda's, not
```

File: src/black/trans.py [hunk 2]
```diff
@@ -1841,22 +1843,23 @@ def _assign_match(LL: List[Leaf]) -> Optional[int]:
...
-            statement requirements listed in the 'Requirements' section of this
-            classes' docstring.
+            statement or lambda expression requirements listed in the
+            'Requirements' section of this classes' docstring.
                 OR
             None, otherwise.
         """
-        # If this line is apart of a dictionary key assignment...
-        if syms.dictsetmaker in [parent_type(LL[0]), parent_type(LL[0].parent)]:
+        # If this line is apart of a dictionary key assignment or lambda expression...
```
Comment (JelleZijlstra) [inline src/black/trans.py:1856]:
```suggestion
        # If this line is a part of a dictionary key assignment or lambda expression...
```

**Review by JelleZijlstra — COMMENTED** [review]
thanks, just noticed two typos

**Comment by yilei** [comment]
Thanks for the review! Agreed above two examples are out of scope from this PR, they concern the dict _keys_. Opened https://github.com/psf/black/issues/3442 separately.

**Review by JelleZijlstra — APPROVED** [review]

**Pull request merged**